core/chain-configs/src/genesis_config.rs

```diff
@@ -266,6 +286,93 @@ impl GenesisRecords {
     }
 }
 
+impl Default for GenesisRecordsFile {
+    fn default() -> Self {
+        GenesisRecordsFile { path: PathBuf::new(), file_type: GenesisRecordsFileType::FullGenesis }
+    }
+}
+
+/// Visitor for records
+/// Reads records one by one and passes them to sink
+struct RecordsProcessor<F> {
+    sink: F,
+}
+
+impl<'de, F: FnMut(StateRecord)> Visitor<'de> for RecordsProcessor<&'_ mut F> {
+    type Value = ();
+
+    fn expecting(&self, formatter: &mut fmt::Formatter) -> fmt::Result {
+        formatter.write_str("array of StateRecord")
+    }
+
+    fn visit_seq<A>(self, mut seq: A) -> Result<Self::Value, A::Error>
+    where
+        A: SeqAccess<'de>,
+    {
+        while let Some(record) = seq.next_element::<StateRecord>()? {
+            (self.sink)(record)
+        }
+        Ok(())
+    }
+}
+
+impl<'de, F: FnMut(StateRecord)> DeserializeSeed<'de> for RecordsProcessor<&'_ mut F> {
+    type Value = ();
+
+    fn deserialize<D>(self, deserializer: D) -> Result<Self::Value, D::Error>
+    where
+        D: Deserializer<'de>,
+    {
+        deserializer.deserialize_seq(self)
+    }
+}
+
+/// Visitor for Genesis
+/// Processes records field with records_processor
+/// IGNORES OTHER FIELDS
+struct GenesisRecordsProcessor<F> {
+    records_processor: RecordsProcessor<F>,
+}
+
+impl<'de, F: FnMut(StateRecord)> Visitor<'de> for GenesisRecordsProcessor<&'_ mut F> {
+    type Value = ();
+
+    fn expecting(&self, formatter: &mut fmt::Formatter) -> fmt::Result {
+        formatter.write_str("map with records field which is array of StateRecord")
+    }
+
+    fn visit_map<A>(self, mut map: A) -> Result<Self::Value, A::Error>
+    where
+        A: MapAccess<'de>,
+    {
+        while let Some(key) = map.next_key::<String>()? {
+            match key.as_str() {
+                "records" => {
+                    map.next_value_seed(self.records_processor)?;
+                    return Ok(());
+                }
+                _ => {
+                    map.next_value::<IgnoredAny>()?;
+                }
+            }
+        }
+        Err(de::Error::custom("missing field: records"))
+    }
+}
```

Comment:
I wonder if it makes sense to collapse these two to a single visitor, which can process either a map or a seq? Just so that we don't have two things and two public functions?

Reply:
I wasn't able to do it when I started writing this code, but now I tried again and everything worked. It's much simpler and prettier this way.